keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1349 @@
+https://storage.googleapis.com/k8s-triage/index.html
+
+- <test>: <link to test coverage>
+-->
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+- <test>: <link to test coverage
+
+-->
+The following scenario will be covered with e2e tests:
+- early job termination when a container fails with a non-retriable exit code
+
+More e2e test scenarios might be considered during implementation if practical.
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, [feature gate] graduations, or as
+something else. The KEP should keep this high-level with a focus on what
+signals will be looked at to determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Feature gate][feature gate] lifecycle
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+-->
+#### Alpha
+
+- Implementation:
+  - handling of failed pods with respect to `spec.backoffPolicy` by Job controller
+  - appending of a dedicated Pod condition (when the Pod termination is
+    initiated by a kubernetes control plane component) to the list of Pod
+    conditions along with sending the Pod delete request
+  - define as a constant and document the new Pod condition Type
+  - the feature is limited by disallowing of the use of `onExitCodes` when
+    `restartPolicy=OnFailure`
+- The feature flag disabled by default
+- Tests: unit and integration
+
+#### Beta
+
+- Address reviews and bug reports from Alpha users
+- E2e tests are in Testgrid and linked in KEP
+- A scalability test to demonstrate the limited impact of the additional API call
+  when terminating a Pod
+- Re-evaluate modification to kubelet to send a dedicated condition when
+  terminating a Pod
```

Comment:
```suggestion
  terminating a Pod, based on user feedback.
```

Maybe we can be a bit more intentional here. We hope to clarify from user feedback, which can be obtained by creating a form and bringing the question to wg-batch and sig-node